core/tests/protractor/libraryFlow.js

```diff
@@ -0,0 +1,215 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview End-to-end tests for library flow.
+ */
+
+var AdminPage = require('../protractor_utils/AdminPage.js');
+var CollectionEditorPage =
+  require('../protractor_utils/CollectionEditorPage.js');
+var ExplorationPlayerPage = require(
+  '../protractor_utils/ExplorationPlayerPage.js');
+var LibraryPage = require('../protractor_utils/LibraryPage.js');
+var general = require('../protractor_utils/general.js');
+var users = require('../protractor_utils/users.js');
+var waitFor = require('../protractor_utils/waitFor.js');
+var workflow = require('../protractor_utils/workflow.js');
+
+describe('Library pages tour', function() {
+  var EXPLORATION_TITLE = 'Test Exploration';
+  var EXPLORATION_OBJECTIVE = 'To learn testing';
+  var EXPLORATION_CATEGORY = 'Algorithms';
+  var EXPLORATION_LANGUAGE = 'English';
+  var EXPLORATION_RATING = 4;
+  var SEARCH_TERM = 'python';
+  var libraryPage = null;
+  var explorationPlayerPage = null;
+
+  beforeEach(function() {
+    libraryPage = new LibraryPage.LibraryPage();
+    explorationPlayerPage = new ExplorationPlayerPage.ExplorationPlayerPage();
+  });
+
+  var visitRecentlyPublishedPage = function() {
+    browser.get('library/recently_published');
+    waitFor.pageToFullyLoad();
+  };
+
+  var rateExploration = function() {
+    if (browser.isMobile) {
+      var adminPage = new AdminPage.AdminPage();
```

Comment:
Here we use a different function to login for the mobile: users.createAndLoginAdminUserMobile. Therefore, the difference in lines. We can however use the same function for desktop too. How should we go about this then? Shall we rename it to something like: createAndLoginAdminUserViaAdminCheckboxOnly or something on that lines?

Reply:
I guess I don't understand why the user is admin in one case and non-admin in the other case, or why the usernames and email addresses used are different. Can we make them the same? Ideally you would not need any branching at all except for reloadExploration() vs createAndPublishExploration().

Reply:
Okay, so we can have a single admin user and then reload or create and publish the exploration but we should use users.createAndLoginAdminUserMobile function instead of users.createAndLoginAdminUser to prevent any popups of the admin page and rename the function appropriately. Does that sounds good? Thanks!

Reply:
So, I didn't actually understand the comment in the code. What popup exactly is appearing?

Reply:
(What does it look like?)

Reply:
My bad, I was referring to the "Roles" tab in the admin page. We had discussed this earlier in one of the review comments [here](https://github.com/oppia/oppia/pull/5340#discussion_r205873285). Basically, for desktop, to login as an admin, we first login with the checkbox "Sign in as admin" ticked and then visit the admin page to update the role of the corresponding username. For mobile, we do the same except for visiting the admin page to update the role of the user since the update role dropdown hides the "Update Role" button and the test times out, something like this:

![localhost_8181_admin pixel 2 1](https://user-images.githubusercontent.com/24826041/43631981-68a8b6c6-9722-11e8-8c0f-0765296d3139.png)
(This has been taken from the Chrome dev tools for Pixel 2. Similar thing happens on Browserstack).

Therefore, I had created a separate function for this purpose as suggested.

However, we can use the same function for desktop too and compress the tests to a single line here as you had suggested.
Thanks!

Reply:
Ohh. Do the same thing for both cases, definitely. One important point to know is that the desktop login procedure you described logs you in as a _super-admin_, not an admin. The Roles flow logs you in as an admin.

Do you even need to log in as an admin though?

Reply:
Actually, I am loading an exploration from the admin page for mobile but for desktop, we are creating a similar exploration which does not require admin access. Therefore, when we unify the test here, we need to login as an admin since the mobile test requires access to the admin page.

Reply:
OK. For clarity, think of the admin page as the "superadmin page".

I get that, on mobile, stuff is loaded from the superadmin page. But I think you solved that in all the other tests already quite cleanly, so I'm not sure why this test is special in that regard...?

Reply:
Okay, understood. This was the approach I had adopted:
```javascript
    users.createUser('random@gmail.com', 'random');
    users.login('random@gmail.com', true);
    if (browser.isMobile) {
      var adminPage = new AdminPage.AdminPage();
      adminPage.reloadExploration('protractor_mobile_test_exploration.yaml');
    } else {
      workflow.createAndPublishExploration(
        EXPLORATION_TITLE,
        EXPLORATION_CATEGORY,
        EXPLORATION_OBJECTIVE,
        EXPLORATION_LANGUAGE
      );
    }
   ...
   ...
   users.logout();

```

Reply:
This seems reasonable :) Try to get the stuff within each branch to a single call, though.

Reply:
Sounds good. I will move the admin page initialisation to outside the branching so as to reduce this to a single call. Thanks :)

Reply:
Done, thanks!